single_kernel_mongo/core/operator.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract Operator for Mongo Related Charms."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+from ops.charm import RelationDepartedEvent
+from ops.framework import Object
+from ops.model import Unit
+
+from single_kernel_mongo.config.literals import RoleEnum, Substrates
+from single_kernel_mongo.managers.config import CommonConfigManager
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm
+    from single_kernel_mongo.core.workload import WorkloadBase
+    from single_kernel_mongo.events.database import DatabaseEventsHandler
+    from single_kernel_mongo.events.tls import TLSEventsHandler
+    from single_kernel_mongo.managers.tls import TLSManager
+
+
+class OperatorProtocol(ABC, Object):
+    """Protocol for a charm operator.
+
+    A Charm Operator must define the following elements:
+     * charm: The Charm it is bound to.
+     * name: The charm operator name, which is one value of the `RoleEnum`
+        enum. This is a class var defined in the operator.
+     * tls_manager: The TLS manager for the mandatory tls events and handlers
+     * state : The CharmState, object handling peer databag interactions, and model interactions.
+     * mongo_manager: The manager for MongoD related interactions.
+     * workload: The main workload of this Charm.
+    """
+
+    charm: AbstractMongoCharm
+    name: ClassVar[RoleEnum]
+    substrate: Substrates
+    config_manager: CommonConfigManager
+    tls_manager: TLSManager
+    state: CharmState
+    mongo_manager: MongoManager
+    workload: WorkloadBase
+    client_events: DatabaseEventsHandler
+    tls_events: TLSEventsHandler
+
+    if TYPE_CHECKING:
+
+        def __init__(self, dependent: AbstractMongoCharm): ...
+
+    @abstractmethod
+    def on_install(self) -> None:
+        """Handles the install event."""
+        ...
+
+    @abstractmethod
+    def on_start(self) -> None:
+        """Handles the start event."""
+        ...
+
+    @abstractmethod
+    def on_secret_changed(self, secret_label: str, secret_id: str) -> None:
+        """Handles the secret changed events."""
+
+    @abstractmethod
+    def on_config_changed(self) -> None:
+        """Handles the config changed events."""
+        ...
```

Comment:
It's a syntax called `Ellipsis` that is used mainly when defining interfaces, and it's also used in another cased (iterator splicing).
In this case, it acts as a no-op which says "This function will be defined elsewhere".
I find it better than providing an implementation (like `raise NotImplementedError`) in fully abstract classes because it leaves the function fully unimplemented which means that all child classes will have to define it or the class cannot be instanced (so early error, at type checking).

Reply:
Thank you so much for the explaination! I really enjoy learning from you!